### PR TITLE
[Diloco] Diloco trainer

### PR DESCRIPTION
**Pull request by khatwanimohit**
# Description

This PR introduces support for Distributed Low-Communication (DiLoCo) training in MaxText. It implements both standard DiLoCo, enabling efficient model training across disjoint clusters ("islands") by synchronizing gradients infrequently via an outer optimizer.

Key Changes
   * Core Logic: Added src/MaxText/diloco.py, which implements the DiLoCoTrainState, inner/outer optimization steps,
     and communication synchronization using drjax.
   * Training Loop Integration: Modified src/MaxText/train.py to initialize the DiLoCo state and adapt the training
     step when enable_diloco is active. This includes handling data reshaping for multiple replicas.
   * Sharding & Configuration:
       * Updated src/MaxText/sharding.py to support a hierarchical "diloco" sharding axis.
       * Added new flags (e.g., enable_diloco, num_diloco_replicas, diloco_outer_optimizer) to base.yml and types.py.
   * Dependencies: Added drjax to the project requirements.
   * Testing: Added comprehensive unit tests in tests/diloco_test.py.

*Notice 1:* Once all tests pass, the "pull ready" label will automatically be assigned.
This label is used for administrative purposes. Please do not add it manually.

*Notice 2:* For external contributions, our settings currently require an approval from a MaxText maintainer to trigger CI tests.

# Tests

Please describe how you tested this change, and include any instructions and/or
commands to reproduce.

# Checklist

Before submitting this PR, please make sure (put X in square brackets):
- [x] I have performed a self-review of my code. For an optional AI review, add the `gemini-review` label.
- [x] I have necessary comments in my code, particularly in hard-to-understand areas.
- [x] I have run end-to-end tests tests and provided workload links above if applicable.
- [x] I have made or will make corresponding changes to the doc if needed, including adding new documentation pages to the relevant Table of Contents (toctree directive) as explained in [our documentation](https://maxtext.readthedocs.io/en/latest/development.html#adding-new-documentation-files).

### CODE REVIEW

File: tests/unit/diloco_test.py [hunk 1]
```diff
@@ -0,0 +1,290 @@
...
+import pytest
+
+from MaxText import diloco
+from MaxText.pyconfig import initialize_pydantic
+from MaxText.globals import MAXTEXT_REPO_ROOT
+
+_BASE_CONFIG_PATH = os.path.join(MAXTEXT_REPO_ROOT, "src", "MaxText", "configs", "base.yml")
+
+
+class SimpleNNXModel(nnx.Module):
```
Comment (khatwanimohit) [inline tests/unit/diloco_test.py:38]: use SimpleLayer

File: src/maxtext/trainers/diloco/diloco.py [hunk 2]
```diff
@@ -0,0 +1,201 @@
...
+
+from MaxText import pyconfig
+
+Batch = Any
+Params = PyTree
+Metrics = PyTree
+OptState = optax.OptState
+InnerOptStates = optax.OptState
+PRNGKey = Key[Array, ""] | UInt32[Array, "2"]
+Step = Int32[Array, ""]
```
Comment (NuojCheng) [inline src/maxtext/trainers/diloco/diloco.py:44]: nit: maybe move to `MaxText.commons`

File: src/maxtext/trainers/diloco/diloco.py [hunk 3]
```diff
@@ -0,0 +1,201 @@
...
+  Args:
+    config: The config used to set up training.
+    train_step: A local train step. This will be executed independently within
+      each replica.
+  """
+  outer_optimizer = optax.sgd(
+      config.diloco_outer_lr,
+      momentum=config.diloco_outer_momentum,
+      nesterov=True,
+  )
```
Comment (NuojCheng) [inline src/maxtext/trainers/diloco/diloco.py:232]: seems outer_optimizer is built twice in line 108 and here. Can we merge them?

File: src/MaxText/train.py [hunk 4]
```diff
@@ -410,17 +411,59 @@ def train_loop(config, recorder, state=None):
...
+        config,
+        model,
+        mesh,
+        state,
+        state_mesh_shardings,
+        train_step,
+        eval_step,
+        eval_data_iterator,
+        params_shardings,
+    )
```
Comment (NuojCheng) [inline src/MaxText/train.py]: maybe move logics out of train.py

File: src/MaxText/train.py [hunk 5]
```diff
@@ -445,6 +491,8 @@ def train_loop(config, recorder, state=None):
 
       with jax.profiler.StepTraceAnnotation("train", step_num=step):
         example_batch = data_loader.load_next_batch(rampup_manager=rampup_manager)
+        if config.enable_diloco:
+          example_batch = diloco.reshape_first_axis_with_diloco(config.num_diloco_replicas, example_batch)
```
Comment (NuojCheng) [inline src/MaxText/train.py]: actually you could consider move this logic and the `config.input_data_sharding_logical_axes` change in `sharding.py` to `MaxText.data_loader`, e.g.

```
def load_next_batch(**args):
   if enable_diloco:
      ....
  else:
    original logics...
```

Comment (NuojCheng) [reply src/MaxText/train.py]: in this case, along with the previous suggested change, you don't need to change anything in `train.py`

File: tests/diloco_test.py [hunk 6]
```diff
@@ -0,0 +1,290 @@
...
+    # test_config = pyconfig.HyperParameters(
+    #     pydantic_config=_TestConfig(
+    #         keys={
+    #             "num_diloco_replicas": num_replicas,
+    #             "diloco_outer_momentum": 0.9,
+    #             "diloco_outer_lr": 1.0,
+    #             "diloco_sync_period": num_steps - 1,
+    #         }
+    #     )
+    # )
```
Comment (NuojCheng) [inline tests/diloco_test.py]: remove

File: tests/unit/diloco_test.py [hunk 7]
```diff
@@ -0,0 +1,290 @@
...
+            _BASE_CONFIG_PATH,
+            f"dcn_diloco_parallelism={num_replicas}",
+            "ici_diloco_parallelism=1",
+            "diloco_outer_momentum=0.9",
+            "diloco_outer_lr=1.0",
+            f"diloco_sync_period={num_steps-1}",
+        ]
+    )
+
+    with mesh:
```
Comment (NuojCheng) [inline tests/unit/diloco_test.py:81]: with jax.set_mesh(mesh), nn_partitioning.axis_rules(config.logical_axis_rules): ...

`with mesh:` context management gonna be deprecated in JAX

File: src/MaxText/train.py (hunk 5 above)
Comment (khatwanimohit) [reply src/MaxText/train.py]: I was making this change I realized we are calling `sharding.maybe_shard_with_name` twice. 
first inside in `data_loader.load_next_batch` and the secondly after `data_loader.load_next_batch` is called in train.py

@NuojCheng can you double check if this is true and then I can remove one of them along with this

Comment (NuojCheng) [reply src/MaxText/train.py]: Yes please remove the one in train.py. Thanks!

Comment (NuojCheng) [reply src/MaxText/train.py]: I have made this change in #2926

File: src/MaxText/train.py (hunk 4 above)
Comment (khatwanimohit) [reply src/MaxText/train.py]: done

File: src/maxtext/utils/train_utils.py [hunk 8]
```diff
@@ -147,6 +157,10 @@ def jit_train_and_eval_step(
     params_shardings=None,
 ):
   """Returns a JIT-compiled train and eval step function."""
+  if config.enable_diloco:
+    # init_diloco_state_partial = functools.partial(diloco.build_diloco_state, config, lambda: state)
```
Comment (gobbleturk) [inline src/maxtext/utils/train_utils.py]: can you remove this comment or clarify it?

Comment (khatwanimohit) [reply src/maxtext/utils/train_utils.py]: removed
